Collect artifacts created on every run worflow step

saraycp opened the pull request:
On every webhook OBS receives from an SCM, OBS follows the instructions given by `workflows.yml` performing the steps on it.

Those steps can handle or create different bunches of artifacts. For example, _repositories_ and _architectures_ if the step is `configure_repositories` or _projects_ and _packages_ in steps like `link_package` etc.

After each workflow run, we want to display all the artifacts involved so the user can better understand what was going on.

In this PR, we are collecting the main information about the artifacts we get on every step and storing them in DB with JSON format.

The JSON for a _link_package_ step can look like this:

```
{
  "source_project": "<source project name>",
  "source_package": "<source package name>",
  "target_project": "<target project name>",
  "target_package": "<target package name>"
}
```

With the information in JSON, we can later display a text and even a link to the artifact from the WorkflowRun show view. 

The view should look like this:

![image (1)](https://user-images.githubusercontent.com/2581944/151948231-d29e92c1-2f87-4a85-8833-3e5d0db7c98e.png)

hennevogel commented on src/api/app/models/workflow/step/branch_package_step.rb:
```diff
@@ -64,4 +66,15 @@ def create_branched_package
+    report = workflow_run.report ? JSON.parse(workflow_run.report) : {}
+    report[step_key] ||= []
+    report[step_key] << { 'source_project' => source_project_name, 'source_package' => source_package_name, 'target_project' => target_project_name, 'target_package' => target_package_name }
+    workflow_run.update(report: report.to_json)
```
If you store json, why don't you [serialize](https://api.rubyonrails.org/classes/ActiveRecord/AttributeMethods/Serialization/ClassMethods.html) it?

saraycp replied:
Sorry, I stashed that change by mistake.

saraycp replied:
Included in my last push now: https://github.com/openSUSE/open-build-service/compare/b73b7db0ad4b62124223778927f89056fced568e..428c6fb5e0ab662dac76b84dfe6d5893da46d0d8

hennevogel commented on src/api/app/models/workflow/step/configure_repositories.rb:
```diff
@@ -64,4 +68,15 @@ def validate_project_name
   end
+
+  def write_report
+    return unless workflow_run = WorkflowRun.find(workflow_run_id)
```
Looks to me like you want to dry this up in `Workflow::Step`. The only difference between the step types is the json content right?

Would also greatly simplify the specs you need to touch etc.

saraycp replied:
I did some refactoring already. I moved the `write_report` to the Step model and in every specific step, I simply define the method `artifacts` that returns only the hash with the information.

danidoni commented on src/api/app/models/workflow/step/branch_package_step.rb:
```diff
@@ -64,4 +66,15 @@ def create_branched_package
+    report = workflow_run.report ? JSON.parse(workflow_run.report) : {}
+    report[step_key] ||= []
+    report[step_key] << { 'source_project' => source_project_name, 'source_package' => source_package_name, 'target_project' => target_project_name, 'target_package' => target_package_name }
+    workflow_run.update(report: report.to_json)
```
I would change `report` to `artifacts`

saraycp replied:
Yes, why not. Let me finish everything using "report" and then I'll replace all the code.

danidoni commented on src/api/db/schema.rb:
```diff
@@ -1066,6 +1066,7 @@
     t.datetime "updated_at", precision: 6, null: false
     t.integer "token_id", null: false
     t.string "response_url"
+    t.text "report"
```
This would be "artifacts", right?

saraycp replied:
could be, yes

hennevogel commented on src/api/app/models/workflow/step.rb:
```diff
@@ -93,8 +93,25 @@ def target_package_name(short_commit_sha: false)
   end
 
+  def write_report
+    return unless (workflow_run = WorkflowRun.find(workflow_run_id))
```
```suggestion
    workflow_run = WorkflowRun.find(workflow_run_id)
    return unless workflow_run
```

hennevogel replied:
Don't make me think :)

saraycp replied:
changed

hennevogel commented on src/api/spec/models/workflow/step/rebuild_package_step_spec.rb:
```diff
@@ -21,10 +22,19 @@
                    })
   end
 
+  let(:written_report) do
```
No need for this for every `::Step` variant right?

saraycp commented:
@danidoni @hennevogel, I have renamed `report` to `artifacts`.
I've also removed the commits related to each specific `::Step`, I'll try to test the artifacts only in `spec/models/workflow/step_spec.rb`

dmarcoux commented on src/api/app/models/workflow_run.rb:
```diff
@@ -2,6 +2,8 @@ class WorkflowRun < ApplicationRecord
   validates :response_url, length: { maximum: 255 }
   validates :request_headers, :status, presence: true
 
+  serialize :report, JSON
```
This should be `artifacts`, right?

hennevogel commented on src/api/app/services/workflows/artifacts_collector.rb:
```diff
@@ -0,0 +1,49 @@
+                    { step.project, step.repositories, step.architectures }
+                  when 'rebuild_package'
+                    { step.project, step.package }
+                  end
```
How do you know what is what? By position in the json?

dmarcoux replied:
For the UI, there will be a view component which knows how to handle every kind of step. With a similar case statement, it will display the artifacts correctly by knowing which step has which keys.

hennevogel replied:
I see, this is just pseudo code...

hennevogel commented on src/api/app/models/workflow_artifacts_per_step.rb:
```diff
@@ -0,0 +1,21 @@
+
+# == Schema Information
+#
+# Table name: workflow_artifacts_per_steps
```
Not quite sure why you would use that implementation detail (that you show them by step) into the table name.

saraycp replied:
The naming is a problem here.

The model, let's call it X, represents a bunch of artifacts related to one step in a JSON format  (i.e. target_package, repositories etc.)

**WorkflowArtifacts**

The first idea was to call the method `WorkflowArtifacts` plural, because is the name that better represents what I described before. Choosing this method name implies not following rails standards and changing the **inflector**.

**WorkflowArtifact**

An alternative is just to call it `WorkflowArtifact` singular. But is not so good semantically. Because one instance of that model doesn't represent one artifact but many (related to one step).

Moreover, having something like `WorkflowArtifact#artifacts` sounds weird.

**WorkflowArtifactsPerStep**

That's why I chose `WorkflowArtifactsPerStep` which semantically is perfect, follows the standard of being singular, and `WorkflowArtifactsPerStep#artifacts` sounds better.

Please let me know if any of you have a better name for the model or for the column `X#artifacts`.

hennevogel replied:
I see, thanks for the explanation.

dmarcoux commented on src/api/app/models/workflow.rb:
```diff
@@ -31,11 +31,17 @@ def call
     when scm_webhook.new_pull_request?, scm_webhook.updated_pull_request?, scm_webhook.push_event?, scm_webhook.tag_push_event?
       steps.each do |step|
-        step.call({ workflow_filters: filters })
+        # TODO: improve this call.
```
What should be improved exactly? Will it happen in this PR?

saraycp replied:
Sorry, I forgot to remove the comment (on it right now). It was added before I moved the code to the method `call_step_and_collect_artifacts` because of high cyclomatic complexity.

saraycp replied:
comment removed

dmarcoux commented on src/api/app/services/workflows/artifacts_collector.rb:
```diff
@@ -0,0 +1,40 @@
+                      package: @step.step_instructions[:package]
+                    }
+                  when 'Workflow::Step::ConfigureRepositories'
+                    # We only need the repositories' names and their archtectures.
```
```suggestion
                    # We only need the repositories' names and their architectures.
```

saraycp replied:
corrected

dmarcoux commented on src/api/app/services/workflows/artifacts_collector.rb:
```diff
@@ -0,0 +1,40 @@
+                  when 'Workflow::Step::ConfigureRepositories'
+                    # We only need the repositories' names and their archtectures.
+                    repositories_and_architectures = @step.step_instructions[:repositories].map do |repository|
+                      repository.delete(:target_project)
```
With the changes from #12128, this needs to be updated:

```suggestion
                      repository.delete(:paths)
```

saraycp replied:
thanks! done

dmarcoux commented on src/api/app/services/workflows/artifacts_collector.rb:
```diff
@@ -0,0 +1,40 @@
+                      target_package: @step.target_package_name
+                    }
+                  when 'Workflow::Step::RebuildPackage'
+                    {
```
This could also be written as `@step.step_instructions.slice(:project, :package)`.

saraycp replied:
good catch!

saraycp commented:
@dmarcoux thanks for your feedback. I did

- rebase on master
- adapt code and specs to deal with the new `paths:` for repositories,
- change/removed/added some comments to the code

saraycp commented on src/api/app/services/workflows/artifacts_collector.rb:
```diff
@@ -0,0 +1,35 @@
+                  when 'Workflow::Step::ConfigureRepositories'
+                    # We only need the repositories' names and their architectures.
+                    repositories_and_architectures = @step.step_instructions[:repositories].map do |repository|
+                      repository.delete(:paths)
```
I'm not so sure the paths are so important :thinking:  but it doesn't harm to display them as we display the architectures.

I'm working on the UI right now. I'll tackle the paths as an improvement as soon as the UI is ready for review.

dmarcoux replied:
When building a package, its dependencies are fetched from the package's project and its repositories. If a repository doesn't have a repository path, it's not providing anything. So yes, repository paths are important!

saraycp replied:
I meant that I wasn't sure that they are important to be in the artifacts, as we are going to provide a link to the repositories tab of the project, where the user can see all the details about them.
But yes, as I said, it's not a problem for me to add them to the artifacts as well.

saraycp replied:
@dmarcoux The paths are already included in the artifacts. Please have a look to the last update of this PR.
